lib/Twig/Environment.php

```diff
@@ -472,7 +473,21 @@ public function loadTemplate($name, $index = null)
             $this->initRuntime();
         }
 
-        return $this->loadedTemplates[$cls] = new $cls($this);
+        if (isset($this->loading[$cls])) {
+            throw new Twig_Error_Runtime(sprintf('Circular reference detected for Twig template "%s", path: %s', $name, implode(' -> ', array_merge($this->loading, array($name)))));
+        }
+
+        $this->loading[$cls] = $name;
+
+        try {
+            $this->loadedTemplates[$cls] = new $cls($this);
+        } catch (\Exception $e) {
+            unset($this->loading[$cls]);
```

Comment:
you need to unset at some point when the template is loaded for real, not just in case of an error, right?